libraries/AP_HAL_Linux/TCPClientDevice.cpp

```diff
@@ -0,0 +1,65 @@
+#include <AP_HAL.h>
+
+#if CONFIG_HAL_BOARD == HAL_BOARD_LINUX
+
+#include <stdio.h>
+#include <unistd.h>
+#include <errno.h>
+#include <stdlib.h>
+
+#include "TCPClientDevice.h"
+
+TCPClientDevice::TCPClientDevice(const char *ip, uint16_t port):
```

Comment:
the TCPClientDevice class doesn't make much sense to me. The "wait" argument isn't intended to switch between a client and a server, it is intended to change the behaviour of a TCP server so that it either waits for a connection or doesn't wait (in which case the accept() happens later when the client connects).
The use case is where you have multiple TCP ports, and you only want to wait for the first one. If you have a look at how HAL_SITL does this you'll see that the first 'A' port waits for a connection, whereas the other ports don't. That allows ardupilot to wait for the first GCS connection before starting (so the GCS sees the startup messages), but still allows the GCS to connect/disconnect to the other ports multiple times during the flight. Try it with SITL to see what I mean, using tcp:127.0.0.1:5760 and tcp:127.0.0.1:5762.
I can't think of a use case for a TCP client device at the moment